test/test_distributions.py

```diff
@@ -522,6 +530,63 @@ def test_gamma_sample_grad(self):
                                        'rel error {}'.format(rel_error),
                                        'max error {}'.format(rel_error.max())]))
 
+    # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_shape(self):
+        df = Variable(torch.exp(torch.randn(2, 3)), requires_grad=True)
+        df_1d = Variable(torch.exp(torch.randn(1)), requires_grad=True)
+        self.assertEqual(Chi2(df).sample().size(), (2, 3))
+        self.assertEqual(Chi2(df).sample_n(5).size(), (5, 2, 3))
+        self.assertEqual(Chi2(df_1d).sample_n(1).size(), (1, 1))
+        self.assertEqual(Chi2(df_1d).sample().size(), (1,))
+        self.assertEqual(Chi2(0.5).sample().size(), (1,))
+        self.assertEqual(Chi2(0.5).sample_n(1).size(), (1,))
+
+        def ref_log_prob(idx, x, log_prob):
+            d = df.data.view(-1)[idx]
+            expected = scipy.stats.chi2.logpdf(x, d)
+            self.assertAlmostEqual(log_prob, expected, places=3)
+
+        self._check_log_prob(Chi2(df), ref_log_prob)
+
+    # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_sample(self):
+        set_rng_seed(0)
+        for df in [0.1, 1.0, 5.0]:
+            self._check_sampler_sampler(Chi2(df),
+                                        scipy.stats.chi2(df),
+                                        'Chi2(df={})'.format(df))
+
+        # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_sample_grad(self):
+        set_rng_seed(1)
+        num_samples = 100
+        for df in [1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4]:
+            # NOTE: for some reason test fails for df = 1e-2
```

Comment:
I'm a bit worried about this failure. Any ideas what might be causing it? Numerical instability? Unlucky seed?

Reply:
I suspect this is due to underflow in the underlying `Gamma` sampler: when `alpha<=0.01` lots of mass ends up at zero (or near zero #4262). I'm even surprised the 1e-3 case passes. For example [these plots](https://github.com/fritzo/notebooks/blob/master/debug-gamma-sampler.ipynb) show the difference between the double-precision scipy sampler and single-precision `torch.distributions.Gamma` sampler. Maybe this test will pass if you use `DoubleTensor` instead of `FloatTensor`?

I doubt this numerical instability it's an issue in practice since `Chi2` is usually created with `df >= 1` (at least in my experience).

Reply:
We're tracking this issue at probtorch#48

Reply:
What worries me is not so much the underflow, as the fact that it passes with 1e-3. But considering your notebook and the fact that it's actually a test that compares to scipy, it might be the case that at values that low they both underflow horribly and compare almost equal 😄 

How about putting up a warning just in case?

Reply:
I just tried df=0.01, it works now. However there's something wrong with the branch I can't build it locally (I have no idea how it passes CI test). I will create a new PR.